R/run_stress_test.R

```diff
@@ -1,55 +1,126 @@
-#' @param discount_rate Numeric, that holds the discount rate of dividends per
+#' @param discount_rate_arg Numeric, that holds the discount rate of dividends per
 #'   year in the DCF. For accepted range compare `discount_rate_range_lookup`.
-#' @param div_netprofit_prop_coef Numeric. A coefficient that determines how
+#' @param div_netprofit_prop_coef_arg Numeric. A coefficient that determines how
 #'   strongly the future dividends propagate to the company value. For accepted
 #'   range compare `div_netprofit_prop_coef_range_lookup`.
-#' @param shock_year Numeric, holding year the shock is applied. For accepted
+#' @param shock_year_arg Numeric, holding year the shock is applied. For accepted
 #'   range compare `shock_year_range_lookup`.
-#' @param term Numeric. A coefficient that determines for which maturity the
+#' @param term_arg Numeric. A coefficient that determines for which maturity the
 #'   expected loss should be calculated in the credit risk section. For accepted
 #'   range compare `term_range_lookup`.
-#' @param company_exclusion Boolean, indicating if companies provided in dataset
+#' @param company_exclusion_arg Boolean, indicating if companies provided in dataset
 #'   excluded_companies.csv shall be excluded.
 #' @return NULL
 #' @export
-run_stress_test <- function(asset_type,
-                            lgd_senior_claims = 0.45,
-                            lgd_subordinated_claims = 0.75,
-                            terminal_value = 0,
-                            risk_free_rate = 0.02,
-                            discount_rate = 0.02,
-                            div_netprofit_prop_coef = 1,
-                            shock_year = 2030,
-                            term = 2,
-                            company_exclusion = TRUE) {
-
-  cat("-- Validating input arguments. \n")
+run_stress_test <- function(asset_type_arg,
+                            lgd_senior_claims_arg = 0.45,
+                            lgd_subordinated_claims_arg = 0.75,
+                            terminal_value_arg = 0,
+                            risk_free_rate_arg = 0.02,
+                            discount_rate_arg = 0.02,
+                            div_netprofit_prop_coef_arg = 1,
+                            shock_year_arg = 2030,
+                            term_arg = 2,
+                            company_exclusion_arg = TRUE) {
+  cat("Running transition risk stress test \n")
+
+  args_list <- mget(names(formals()), sys.frame(sys.nframe()))
+  iter_var <- get_iter_var(args_list)
+  args_tibble <- tibble::as_tibble(args_list)
+
+  st_results_list <- purrr::map(1:nrow(args_tibble), run_stress_test_iteration, args_tibble = args_tibble)
+
+  result_names <- names(st_results_list[[1]])
+  st_results <- result_names %>%
+    purrr::map(function(tib) {
+      purrr::map_dfr(st_results_list, `[[`, tib)
+    }) %>%
+    purrr::set_names(result_names)
+
+  write_stress_test_results(
```

Comment:
This function seems to be doing too much and leave us with a [side-effect soup](https://design.tidyverse.org/side-effect-soup.html). I would split it in at least two functions -- maybe internal:

* One function does the computation and returns an object.
* The other one produces the side-effect -- it writes to disk. 

You may expose a wrapper that does both (although I dislike it) but having those two functions will make testing a lot easier. You generally don't see it if you first write the function and then test it; but you do if you do both in parallel. Code that is easy to test is easy to maintain. 

My ideal API would be this:

```r
"/path/to/data/" %>%
  read_stress_test_data() %>%
  run_stress_test() %>%
  write_stress_test("/path/to/output/directory")
```

Even if this is only internal, it would be a lot easier to maintain.

In general, do complex things by combining simple pieces (like lego) not by creating complex pieces.

If that's not the API you want to expose, you may wrap it in `do_it_all()` or provide a GUI like [this one](https://github.com/2DegreesInvesting/r2dii.climate.stress.test/pull/173). (Note that GUI can't to the iteration as is.)

﻿